### PR TITLE
Feature/281 triangulation covariance

### PR DESCRIPTION
**Pull request by JulianHammerl**
* **Tickets addressed:** Confluence 281
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

The computation of the triangulation covariance matrix was added to the triangulation() method of the cameraTriangulation module. This module now computes the covariance matrix and writes it to the cameraLocalizationMsg. The commits are organized as follows:

- Commit 1: Update necessary message definitions
- Commit 2: Update module
- Commit 3: Update UnitTest
- Commit 4: Update Documentation

## Verification

The UnitTest of the cameraTriangulation module was updated to check if the covariance matrix is computed accurately.

## Documentation

The documentation of the cameraTriangulation module was updated to mention that the covariance is also computed by the module.

## Future work

The triangulation() method may be moved to a library outside the module.

### CODE REVIEW

File: src/architecture/msgPayloadDefC/CameraLocalizationMsgPayload.h [hunk 1]
```diff
@@ -27,6 +27,7 @@ typedef struct {
     uint64_t timeTag; //!< [ns] vehicle time-tag when images where taken
     double sigma_BN[3]; //!< [-] MRP orientation of spacecraft when images where taken
     double cameraPos_N[3]; //!< [m] Camera location in inertial frame */
+    double covariance[3*3]; //!< [m^2] covariance of estimated camera location
```
Comment (thibaudteil) [inline src/architecture/msgPayloadDefC/CameraLocalizationMsgPayload.h]: I would add a _N to clarify the frame

File: src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp [hunk 2]
```diff
@@ -126,9 +130,15 @@ void CameraTriangulation::readMessages()
...
-    this->cameraCalibrationMatrixInverse << 1./dX, -alpha/(dX*dY), (alpha*vp - dY*up)/(dX*dY),
+    Eigen::Matrix3d Kinv;
+    Kinv << 1./dX, -alpha/(dX*dY), (alpha*vp - dY*up)/(dX*dY),
                                             0., 1./dY, -vp/dY,
                                             0., 0., 1.;
+    this->cameraCalibrationMatrixInverse = Kinv;
+
+    Eigen::MatrixXd S(2,3);
+    S << Eigen::Matrix2d::Identity(), Eigen::MatrixXd::Zero(2, 1);
+    this->Rx = Kinv*S.transpose()*this->Ru*S*Kinv.transpose();
```
Comment (thibaudteil) [inline src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp:133]: If this is the same code as in other parts of the module, I'd consider making this a method that can be called in each location

**Review by thibaudteil — COMMENTED** [review]
Nice addition

File: src/architecture/msgPayloadDefC/CameraLocalizationMsgPayload.h (hunk 1 above)
Comment (JulianHammerl) [reply src/architecture/msgPayloadDefC/CameraLocalizationMsgPayload.h]: Done.

File: src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp (hunk 2 above)
Comment (JulianHammerl) [reply src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp:133]: This is the only part of the module with this code. I will mark as resolved.

**Review by thibaudteil — APPROVED** [review]
Already commented on the branch

File: src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp [hunk 3]
```diff
@@ -176,21 +186,24 @@ void CameraTriangulation::writeMessages(uint64_t currentSimNanos)
...
     @param dcmCamera dcm from frame of interest F to camera frame C
-    @return Eigen::Vector3d
+    @return std::pair<Eigen::Vector3d, Eigen::Matrix3d>
 */
-Eigen::Vector3d CameraTriangulation::triangulation(std::vector<Eigen::Vector3d> knownLocations,
-                                                       std::vector<Eigen::Vector2d> imagePoints,
-                                                       const Eigen::Matrix3d& cameraCalibrationInverse,
-                                                       std::vector<Eigen::Matrix3d> dcmCamera) const
+std::pair<Eigen::Vector3d, Eigen::Matrix3d> CameraTriangulation::triangulation(
+        std::vector<Eigen::Vector3d> knownLocations,
```
Comment (patkenneally) [inline src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp]: Weird formatting here.

**Review by patkenneally — APPROVED** [review]

File: src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp (hunk 3 above)
Comment (JulianHammerl) [reply src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp]: fixed it.

**Pull request merged**